app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java

```diff
@@ -598,149 +654,136 @@ public int getFilePosition(final File file) {
-                    if (appDataFolder.exists() || appDataFolder.mkdir()) {
-                        virtualMapping.put(VIRTUAL_STORAGE_APP_DATA_PRIVATE, appDataFolder);
-                        newData.add(VIRTUAL_STORAGE_APP_DATA_PRIVATE);
-                    }
-                } else if (folder.equals(VIRTUAL_STORAGE_RECENTS)) {
-                    newData.addAll(_dopt.recentFiles);
-                } else if (folder.equals(VIRTUAL_STORAGE_POPULAR)) {
-                    newData.addAll(_dopt.popularFiles);
-                } else if (folder.equals(VIRTUAL_STORAGE_FAVOURITE)) {
-                    newData.addAll(_dopt.favouriteFiles);
-                } else if (folder.isDirectory()) {
-                    GsCollectionUtils.addAll(newData, folder.listFiles(GsFileBrowserListAdapter.this));
-                }
+        final File toShow = show == null ? _fileToShowAfterNextLoad : show;
+        _fileToShowAfterNextLoad = null;
 
-                // Some special folders get special children
-                if (folder.getAbsolutePath().equals("/storage/emulated")) {
-                    newData.add(new File(folder, "0"));
-                } else if (folder.getAbsolutePath().equals("/")) {
-                    newData.add(new File(folder, "storage"));
-                } else if (folder.equals(_context.getFilesDir().getParentFile())) {
-                    // Private AppStorage: Allow to access to files directory only
-                    // (don't allow access to internals like shared_preferences & databases)
-                    newData.add(new File(folder, "files"));
-                }
+        executorService.execute(() -> _loadFolder(toLoad, toShow));
+    }
 
-                for (final File externalFileDir : ContextCompat.getExternalFilesDirs(_context, null)) {
-                    for (final File file : newData) {
-                        final String absPath = file.getAbsolutePath();
-                        final String absExt = externalFileDir.getAbsolutePath() ;
-                        if (!canWrite(file) && !absPath.equals("/") && absExt.startsWith(absPath)) {
-                            final int depth = GsTextUtils.countChars(absPath, '/')[0];
-                            if (depth < 3) {
-                                final File parent = file.getParentFile();
-                                if (parent != null) {
-                                    final File remap = new File(parent.getAbsolutePath(), "appdata-public (" + file.getName() + ")");
-                                    virtualMapping.put(remap, new File(absExt));
-                                    newData.add(remap);
-                                }
-                            }
-                        }
-                    }
-                }
+    // This function is not called on the main thread, so post to the UI thread
+    private synchronized void _loadFolder(final @NonNull File folder, final @Nullable File toShow) {
```

Comment:
Breaking the inner function out makes it much cleaner (synchronised in signature)